tensorflow_addons/losses/center_loss.py

```diff
@@ -0,0 +1,86 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements center loss."""
+
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import FloatTensorLike
+from typeguard import typechecked
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class CenterLoss(tf.keras.layers.Layer):
+    """Computes the center loss.
+
+    See: https://ydwen.github.io/papers/WenECCV16.pdf
+    The loss was designed to develop an effectively improve the discriminative
+    power of the deep learned features. In fact, the key is to minimize the
+    intra-class variations while keeping the features of different classes
+    separable. Usually combining with softmax losss to jointly supervise the
+    CNNs to have a better result.
+    We expect labels `y_pred` must be the output of the last fully connected
+    layer, which is 2-D float `Tensor` of l2 normalized embedding vectors and
+    `y_true` to be provided as 1-D integer `Tensor` with shape [batch_size]
+    of multi-class integer labels.
+
+    Args:
+      name: Optional name for the op.
+      alpha: A scalar between 0-1 to control the leanring rate of the centers.
+
+    Returns:
+      Tensor containing calculated center loss.
+    """
+
+    @typechecked
+    def __init__(
+        self, alpha: FloatTensorLike = 0.2, name: str = "center_loss", **kwargs
+    ):
+        super().__init__(name=name, **kwargs)
+        self.alpha = alpha
+
+    def call(self, y_true, y_pred):
+        return self.loss_fn(y_true, y_pred, self.alpha)
+
+    def loss_fn(self, labels, feature, alpha):
+        labels = tf.convert_to_tensor(labels)
+        feature = tf.convert_to_tensor(feature)
+        len_features = feature.shape[-1]
+        num_classes = labels.shape[-1]
+        centers = tf.Variable(
+            tf.constant(0.0, shape=[num_classes, len_features]),
+            name="centers",
+            dtype=tf.float32,
+        )
+        labels = tf.reshape(labels, [-1])
+        centers_batch = tf.gather(centers, labels)
+        diff = centers_batch - feature
+        unique_label, unique_idx, unique_count = tf.unique_with_counts(labels)
+        appear_times = tf.gather(unique_count, unique_idx)
+        appear_times = tf.reshape(appear_times, [-1, 1])
+        diff = diff / tf.cast((1 + appear_times), tf.float32)
+        diff = alpha * diff
+        labels = tf.reshape(labels, [-1, 1])
+        # update centers
+        centers_update_op = tf.tensor_scatter_nd_sub(centers, labels, diff)
```

Comment:
Does the variable needs to be trainable? Also does the variable needs to be re-created at each forward pass?

Reply:
Yes that variable is trainable IMO. Also, with what I understood in the loss, I think the centers are shifting. So, it will be recalculated at each forward pass.